File: 20.py
#20. Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
# An input string is valid if:
#1. Open brackets must be closed by the same type of brackets.
#2. Open brackets must be closed in the correct order.

class Stack:
    def __init__(self, items = None):
        if(items == None):
            self.items = []
        else:
            self.items = items
    def push(self, i):
        self.items.append(i)
    def pop(self):
        return self.items.pop()
    def size(self):
        return len(self.items)
    def isEmpty(self):
        return self.itmes == []
    
def isMatch(open, close) -> bool:
    return (open == '(' and close == ')') or (open == '[' and close == ']') or (open == '{' and close == '}')

def isValid(s: str) -> bool:
    stack = Stack()
    i = 0
    error = 0
    for i in s:
        if(i in '([{'):
            stack.push(i)
        else:
            if(i in ')]}'):
                if(stack.size() > 0):
                    if(not isMatch(stack.pop(), i)):
                        error += 1
                else:
                    error += 1
    if(stack.size() > 0):
        error+=1
    return error == 0

if __name__ == "__main__":
    s = str(input())
    print(isValid(s))

File: golang/20.go
package main

import "fmt"

func main() {
	fmt.Println(isValid("()"))
	fmt.Println(isValid("()[]{}"))
	fmt.Println(isValid("(]"))
}

type Stack []byte

func (s *Stack) isEmpty() bool {
	return len(*s) == 0
}

func (s *Stack) Push(c byte) {
	*s = append(*s, c)
}

func (s *Stack) Pop() byte {
	if len(*s) == 0 {
		return '-'
	} else {
		index := len(*s) - 1
		element := (*s)[index]
		*s = (*s)[:index]
		return element
	}
}

func isValid(s string) bool {
	var stack Stack
	for i := 0; i < len(s); i++ {
		if isInOpen(s[i]) {
			stack.Push(s[i])
		} else {
			if stack.isEmpty() {
				return false
			}
			c := stack.Pop()
			if isMatch(c, s[i]) {
				continue
			} else {
				return false
			}
		}
	}
	if !stack.isEmpty() {
		return false
	}
	return true
}

func isInOpen(c byte) bool {
	s := []byte{'(', '[', '{'}
	for i := 0; i < len(s); i++ {
		if s[i] == c {
			return true
		}
	}
	return false
}

func isMatch(a byte, b byte) bool {
	if a == '(' && b == ')' {
		return true
	} else if a == '[' && b == ']' {
		return true
	} else if a == '{' && b == '}' {
		return true
	}
	return false
}
